File: ruby/lib/ter/runtime/resolver.rb
# frozen_string_literal: true

module Ter
  module Runtime
    # Resolves values from multiple sources with priority ordering
    class Resolver
      SOURCES = {
        process: 3,
        file: 2,
        injected: 1,
        defaults: 0
      }.freeze

      attr_reader :schema

      def initialize(schema)
        @schema = schema
        @sources = {
          process: {},
          file: {},
          injected: {},
          defaults: {}
        }
        @metadata = {}
      end

      # Add process environment variables
      def add_process_env(env = ENV)
        @sources[:process] = env.to_h
        self
      end

      # Add values from file
      def add_file_env(values)
        @sources[:file] = values || {}
        self
      end

      # Add injected values
      def add_injected(values)
        @sources[:injected] = values || {}
        self
      end

      # Resolve all values
      def resolve
        results = {}
        errors = []

        @schema.get_variables.each do |name, type|
          value = find_value(name)

          # Use default if not found
          if value.nil?
            if !type.default_value.nil?
              value = type.default_value
              @metadata[name] = {
                source: 'defaults',
                resolved_at: Time.now.to_i
              }
            end
          end

          result = type.validate(value)

          if result[:valid]
            results[name] = result[:value]
            @metadata[name] ||= {
              source: find_source(name),
              resolved_at: Time.now.to_i
            }
          else
            errors << { field: name, error: result[:error] }
          end
        end

        {
          valid: errors.empty?,
          values: results,
          errors: errors,
          metadata: @metadata
        }
      end

      # Get metadata for a variable
      def get_metadata(name)
        @metadata[name] || {}
      end

      # Get all metadata
      def get_all_metadata
        @metadata.dup
      end

      private

      def find_value(name)
        # Priority: process > file > injected > defaults
        SOURCES.keys.each do |source|
          value = @sources[source][name]
          return value unless value.nil?
        end
        nil
      end

      def find_source(name)
        SOURCES.keys.reverse.each do |source|
          return source.to_s if @sources[source][name]
        end
        'unknown'
      end
    end
  end
end

File: python/ter/runtime/resolver.py
"""Multi-source environment value resolver"""

import os
from typing import Dict, Any, Optional
from dataclasses import dataclass
from datetime import datetime


@dataclass
class ResolvedValue:
    """Value resolved with metadata"""
    value: Any
    source: str  # "process", "file", "injected", "default"
    resolved_at: datetime


class Resolver:
    """Resolves environment values from multiple sources"""
    
    def __init__(self, schema=None):
        self.schema = schema
        self._file_env: Dict[str, str] = {}
        self._injected: Dict[str, str] = {}
    
    def add_file_env(self, env_dict: Dict[str, str]) -> "Resolver":
        """Add environment values from file"""
        self._file_env.update(env_dict)
        return self
    
    def add_injected(self, env_dict: Dict[str, str]) -> "Resolver":
        """Add injected environment values"""
        self._injected.update(env_dict)
        return self
    
    def resolve(self, name: str) -> Optional[ResolvedValue]:
        """Resolve a value from multiple sources"""
        # Try process.env
        if name in os.environ:
            return ResolvedValue(
                value=os.environ[name],
                source="process",
                resolved_at=datetime.now()
            )
        
        # Try file environment
        if name in self._file_env:
            return ResolvedValue(
                value=self._file_env[name],
                source="file",
                resolved_at=datetime.now()
            )
        
        # Try injected
        if name in self._injected:
            return ResolvedValue(
                value=self._injected[name],
                source="injected",
                resolved_at=datetime.now()
            )
        
        # Try default from schema
        if self.schema:
            var_def = self.schema.get_variable(name)
            if var_def and var_def.default:
                return ResolvedValue(
                    value=var_def.default,
                    source="default",
                    resolved_at=datetime.now()
                )
        
        return None
    
    def resolve_all(self) -> Dict[str, ResolvedValue]:
        """Resolve all known variables"""
        resolved = {}
        
        if self.schema:
            for name in self.schema.get_variables():
                value = self.resolve(name)
                if value:
                    resolved[name] = value
        
        return resolved
